scripts/run-live-build.sh

```diff
@@ -129,36 +129,64 @@ while ! curl --output /dev/null --silent --head --fail \
 done
 set -o errexit
 
+pkg_mirror_secondary=''
 if [[ -n "$DELPHIX_PACKAGE_MIRROR_SECONDARY" ]]; then
-	sed "s|@@URL@@|$DELPHIX_PACKAGE_MIRROR_SECONDARY|" \
-		<config/archives/delphix-secondary-mirror.list.in \
-		>config/archives/delphix-secondary-mirror.list
+	pkg_mirror_secondary="$DELPHIX_PACKAGE_MIRROR_SECONDARY"
+else
+	#
+	# If no secondary package mirror is provided, then pull in the latest
+	# mirror dataset for the build. If no latest dataset is found, then fail.
+	#
+	source_url="http://linux-package-mirror.delphix.com/$UPSTREAM_PRODUCT_BRANCH/latest/"
```

Comment:
How will the `UPSTREAM_PRODUCE_BRANCH` be specified?

Reply:
This is set by the environment variable and should be the corresponding Delphix branch, e.g. `master` or `6.0/stage`.

Reply:
@grodr Coming back to the issue of building manually, we should probably use a similar logic in here as we do in the other place that consumes UPSTREAM_PRODUCT_BRANCH: https://github.com/delphix/appliance-build/blob/master/scripts/build-ancillary-repository.sh#L145.

Reply:
> This is set by the environment variable

Where does this environment variable get set?

Reply:
> This is set by the environment variable

Where does this environment variable get set?

re:
```
upstream_branch="${UPSTREAM_PRODUCT_BRANCH:-master}"
```
This code will break once we branch, so I don't think this is something we should duplicate.

Reply:
The alternative is to fail the manual build if this is not provided.

Reply:
@pzakha I'm sure I can work with @grodr to come up with a solution that works for our requirements; which is supporting our branching model, and enables manual builds without having to manually specify this variable.

With that said, I don't know if this has already been solved or not, which is why I was trying to ask @grodr for more details.

Reply:
It is only set in the Jenkins job. So this won't actually work for a developer workflow without that set, I didn't realize that.

Reply:
We are planning on branching appliance-build at the same cadence as the product though right? So could we use the appliance-build branch for this?

Reply:
> We are planning on branching appliance-build at the same cadence as the product though right?

That's my understanding, yes.

> So could we use the appliance-build branch for this?

Given any "local" branch or commit, how would you determine the "product branch" from that? e.g. will this work if we're building for a project branch? or one a clone without all git history and branch information?

That general idea is correct, but I'd have to think more about the implementation of it, to know if it'd work for all cases that might arise.

Reply:
Generally, trying to figure out the product branch from a git tree is not particularly trivial logic. One thing we could do is to keep defaulting to "master", but when a build starts, we should display the variables that are normally set by Jenkins so that the user is aware of what is being used.

Reply:
> One thing we could do is to keep defaulting to "master",

How will this get changed/updated/fixed for each release branch? Are you proposing we use "master" as the default for non-master branches?

> but when a build starts, we should display the variables that are normally set by Jenkins so that the user is aware of what is being used.

How do we get the values that are normally set by Jenkins when building outside of Jenkins?

Reply:
With my proposal, building manually will default to master on non-master branches as well, which will naturally not do the expected thing. One alternative would be to not use any default, and instead put some checks in the topmost `build.gradle` that will fail if some variables are not set and give a suggestion to resolve this. This would mean a few extra steps for building manually but would avoid unexpected behaviour when building older product branches.

Reply:
What if we add a `BRANCH` variable that represents the appliance-build upstream branch. That variable can be updated by the release lead on branching, and pulled in to determine which branch we should pull in from the mirror.

Reply:
> which will naturally not do the expected thing.

If we can't do the correct thing, I'd much rather explicitly fail with an error, than silently do the wrong thing; this sounds like the worst way to handle this issue, IMO.

Reply:
Before suggesting any possible solutions, I need to better understand the scenarios that are most common to occur.

E.g. I wonder if we can have automatic product release branch resolution for most cases (e.g. a "normal" git clone operation when building manually), but then throw an error when that fails, and rely on the user to specific the environment variable correctly on failure.

I might not have time to investigate this immediately, though.

Reply:
I'd propose to go for this approach for now:
> One alternative would be to not use any default, and instead put some checks in the topmost build.gradle that will fail if some variables are not set and give a suggestion to resolve this. This would mean a few extra steps for building manually but would avoid unexpected behaviour when building older product branches.

This should be easy to implement and doesn't cause any unexpected behaviour.

If we want to add more means to automatically determine the product branch, then we can do that later.

Reply:
@pzakha Any reason not to add the default branch as a variable, and have it updated on branching per branch?

Reply:
This is one more manual step that we'd need to do each time we are branching. Given that this variable would not be consumed by the regular code paths (Jenkins will provide it normally), I don't want to impose extra steps into the release process.

Reply:
If we force the user to pass in the correct value, how will we communicate to them which value to use (and when)?

Reply:
"Error, UPSTREAM_PRODUCT_BRANCH is empty. Please provide the upstream branch for the version of Delphix you are building (e.g. master, 6.0/stage) by running 'export UPSTREAM_PRODUCT_BRANCH=<value>".

Reply:
I don't think it's reasonable to assume folks will know what value to use, based only on that error message.

Sure, the folks in the PR thread might know what to set it to, but we're "experts"; I wouldn't expect a user from another team to know, without any other information.

As a design principle, folks from other teams should be able to build this by just looking at the instructions from the README. I also don't think it's good design to communicate build requirements and input parameters in the form of error messages.

Reply:
I think folks should know which version of the product they are building, and you can add this to the README... In any case, I feel like we are overthinking this right now. Manual build is mainly for doing changes to appliance-build, which would most likely involve the platform/devops team anyway. If you guys wanna implement some other ideas, feel free to go ahead with it, I'm not gonna block this. However, if you want to add steps to the release process, make sure to inform the release team.

Reply:
> I think folks should know which version of the product they are building, and you can add this to the README...

The onus to update the README isn't on me, it's on the person making the change to add the new requirement.

> if you want to add steps to the release process

I don't know if you're referring to me specifically here, but to be clear, I never suggested making any changes to the release process.

Reply:
By "you" I was referring to "whoever is working" on this change, not to you personally. In this case, this would be Garret.

Reply:
I will be sure that the release lead is informed of the change. I think adding the BRANCH variable would solve the problems that have been raised.
I could also add something in the README if we think it is necessary. It definitely would be if we require environment variables be set. I am not sure if it needs to be updated if I am just adding the BRANCH variable.

Reply:
> I am not sure if it needs to be updated if I am just adding the BRANCH variable.

If the value has a proper default value, then I don't see any reason for a user to change it, so there's no need to discuss it in the build instruction in the README.

With that said, as an alternative to a hard-coded value, I think it might be reasonable to rely on the "upstream tracking branch" for this information. We already rely on this for some of our "git-utils" commands, so it's generally a requirement we have for other repositories and workflows. Additionally, it should "just work" give the build instructions in the README.

For example, we could use something like:
```
$ git for-each-ref --format='%(upstream:short)' "$(git symbolic-ref -q HEAD)"
```

Here's an example of what this returns for all of our current branches:
```
$ for branch in master staging trying projects/6.0-ea; do
> git clone https://github.com/delphix/appliance-build.git --branch "$branch" &>/dev/null
> cd appliance-build
> git for-each-ref --format='%(upstream:short)' "$(git symbolic-ref -q HEAD)"
> cd ..
> rm -rf appliance-build
> done
origin/master
origin/staging
origin/trying
origin/projects/6.0-ea
```

Obviously this wouldn't work when the "upstream tracking branch" isn't set; but in that case, I'd be fine throwing an error because I doubt folks will actually run into that problem.

Additionally, for the Jenkins case where it might not have a "upstream tracking branch", we'll already be passing that value into the build, rather than relying on this "best effort" logic. I also think this falls squarely into the two-pronged design principle that Basil mentioned before; i.e. Jenkins (or any consumer) can directly specify the variable to use when it knows best, and when it's not specified, we attempt to infer the correct value using the "upstream tracking branch" information.

Would this address all cases, without having to have a hard coded value that needs to be updated for each release branch, and without forcing users to manually specify the value?

Reply:
I'm OK with that. So the logic would be:
1. Check if `UPSTREAM_PRODUCT_BRANCH` is set
2. If not, try to get it from the git upstream tracking branch (making sure to sanitize it properly).
3. If the above fails, ask user to either set upstream branch or to manually set UPSTREAM_PRODUCT_BRANCH.

There's still a chance that the user will set the tracking upstream branch to a branch that isn't a product branch (e.g. some project branch), but that's the best we can do without manually entering the branch. It might be useful to display the value UPSTREAM_PRODUCT_BRANCH when starting the build.

Reply:
This seems fine to me, especially if we require users to set an upstream branch in their set up. I will work on implementing this.

Reply:
After thinking about this a little more, this may not work exactly like I was thinking, since this repository will be on the bootstrap system, rather than on a developer's workstation.

A workflow I use often is to clone my personal git repository onto the bootstrap system, and use a feature branch to build/test. In that example, the upstream tracking branch will be my personal git repository and feature branch name (which won't match our product release branch name). So, for this workflow, I don't think the automatic branch name detection using the upstream tracking branch will work.

With that said, I'm OK with any of the possible designs proposed so far. I'm having a hard time deciding which approach I prefer, since they all come with unique benefits and downsides, so I'm OK with whatever y'all think is best.

Reply:
If that is a common workflow then I lean towards hard coding the `BRANCH` parameter in the repo and having the release lead change the value on branching. I think that will give the smallest chance for a developer having a misconfiguration.

Reply:
This is not the way I'd go for it, but if the release team is OK with taking this extra burden then I'm fine with this. Note that this would mean that when we are branching we need to make 3 PRs for changing this variable in 3 different branches.

Alternatively, if you decide to go for the approach where the user has to provide the upstream product branch when building manually, then we can always switch back to the BRANCH variable hardcoding later on if users of this job script this too cumbersome or confusing.